bindings/java/src/org/sleuthkit/datamodel/HostAddressManager.java

```diff
@@ -308,38 +332,177 @@ private HostAddress getHostAddress(long id, CaseDbConnection connection) throws
+		try (CaseDbConnection connection = this.db.getConnection()) {
+			addHostNameAndIpMapping(dnsNameAddress, ipAddress, time, source, connection);
+		} catch (SQLException ex) {
+			throw new TskCoreException(String.format("Error adding host DNS address mapping for DNS name = %s, and IP address = %s", dnsNameAddress.getAddress(), ipAddress.getAddress()), ex);
+		} finally {
+			db.releaseSingleUserCaseWriteLock();
+		} 
+	}
+	
+	/**
+	 * Adds a row to the host address dns ip map table.
+	 *
+	 * @param dnsNameAddress    The DNS name.
+	 * @param ipAddress         An IP address associated with the DNS name.
+	 * @param time              Timestamp when this relationship was true.
+	 * @param source            The source.
+	 * @param caseDbTransaction The transaction in the scope of which the
+	 *                          operation is to be performed, managed by the
+	 *                          caller. Null is not permitted.
+	 *
+	 * @throws TskCoreException
+	 */
+	public void addHostNameAndIpMapping(HostAddress dnsNameAddress, HostAddress ipAddress, Long time, Content source, final SleuthkitCase.CaseDbTransaction caseDbTransaction) throws TskCoreException {
+
+		if (Objects.isNull(caseDbTransaction)) {
+			throw new IllegalArgumentException("null caseDbTransaction passed to addHostNameAndIpMapping");
+		}
+		try {
+			addHostNameAndIpMapping(dnsNameAddress, ipAddress, time, source, caseDbTransaction.getConnection());
+		} catch (SQLException ex) {
+			throw new TskCoreException(String.format("Error adding host DNS address mapping for DNS name = %s, and IP address = %s", dnsNameAddress.getAddress(), ipAddress.getAddress()), ex);
+		} 
+	}
+	
+	/**
+	 * Adds a row to the host address dns ip map table.
+	 *
+	 * @param dnsNameAddress The DNS name.
+	 * @param ipAddress      An IP address associated with the DNS name.
+	 * @param time           Timestamp when this relationship was true.
+	 * @param source         The source.
+	 * @param connection     The db connection. Null is not permitted.
+	 *
+	 * @throws TskCoreException
+	 */
+	private void addHostNameAndIpMapping(HostAddress dnsNameAddress, HostAddress ipAddress, Long time, Content source, final CaseDbConnection connection) throws  SQLException {
 
 		if (dnsNameAddress.getAddressType() != HostAddress.HostAddressType.HOSTNAME) {
 			throw new IllegalArgumentException("A host name address is expected.");
 		}
 		if ((ipAddress.getAddressType() != HostAddress.HostAddressType.IPV4) && (ipAddress.getAddressType() != HostAddress.HostAddressType.IPV6)) {
 			throw new IllegalArgumentException("An IPv4/IPv6 address is expected.");
 		}
+		if (Objects.isNull(connection)) {
+			throw new IllegalArgumentException("null connection passed to addHostNameAndIpMapping");
+		}
 
 		String insertSQL = db.getInsertOrIgnoreSQL(" INTO tsk_host_address_dns_ip_map(dns_address_id, ip_address_id, source_obj_id, time) "
 				+ " VALUES(?, ?, ?, ?) ");
 
-		db.acquireSingleUserCaseWriteLock();
```

Comment:
I think we should either pass in the transaction (so we know we already have the write lock) or get the write lock (possibly for the second time)

Reply:
@APriestman There are two public variants of this method, one that takes a transaction and another that provides the connection.  Are you suggesting that I reacquire the lock write lock regardless of the manner in which the call gets here?

Reply:
So I think maybe addHostNameAndIpMapping() should not get the lock - I'm under the impression that for these new classes we've gone with acquiring the lock only around the actual SQL call. It'd only be redundant for the transaction case. If you're against that at least make sure there's a prominent comment about having the write lock before calling that method.

Reply:
@APriestman I have moved the lock closer to the statements.